### PR TITLE
[#765]  support Hibernate composite user types

### PR DESCRIPTION
**Pull request by AFbenHassine**
Based on issue: https://github.com/Blazebit/blaze-persistence/issues/765

### CODE REVIEW

File: core/impl/src/main/java/com/blazebit/persistence/impl/EntityMetamodelImpl.java [hunk 1]
```diff
@@ -337,6 +339,20 @@ private TemporaryExtendedManagedType collectColumnNames(EntityType<?> e, Map<Str
...
+                else if (jpaProvider.getJpaMetamodelAccessor().isCompositeUserType(delegate, attribute)) {
+                    Collection<String> compositePropertyNames = jpaProvider.getJpaMetamodelAccessor().getCompositePropertyNames(delegate, attribute);
+
+
+
+                    for (String compositePropertyName : compositePropertyNames) {
+                        String idPath = attributeName + "." + compositePropertyName;
+                        List<Attribute<?,?>> idParents = new ArrayList<>();
+                        idParents.add(attribute);
+                        AttributeEntry attributeEntry = new AttributeEntry(jpaProvider, type, attribute, idPath, BigDecimal.class, idParents, null);
```
Comment (beikov) [inline core/impl/src/main/java/com/blazebit/persistence/impl/EntityMetamodelImpl.java]: We need a way to access the property information like the java type, column type, nullability etc. Here you are always registering `BigDecimal` as type. I know it's just a PoC, I'm just mentioning it.

File: core/api/src/main/java/com/blazebit/persistence/spi/JpaMetamodelAccessor.java [hunk 2]
```diff
@@ -76,4 +77,8 @@
      * @return true if the attribute is an element collection, false otherwise
      */
     boolean isElementCollection(Attribute<?, ?> attribute);
+
+    boolean isCompositeUserType(Metamodel metamodel, Attribute<?,?> attribute);
```
Comment (beikov) [inline core/api/src/main/java/com/blazebit/persistence/spi/JpaMetamodelAccessor.java]: `Metamodel` isn't necessary

File: core/api/src/main/java/com/blazebit/persistence/spi/JpaMetamodelAccessor.java [hunk 3]
```diff
@@ -76,4 +77,8 @@
...
+
+    Collection<String> getCompositePropertyNames(Metamodel metamodel, Attribute<?,?> attribute);
```
Comment (beikov) [inline core/api/src/main/java/com/blazebit/persistence/spi/JpaMetamodelAccessor.java]: `Metamodel` isn't necessary

**Comment by JWGmeligMeyling** [comment]
> Overall it looks interesting, although I think we could try to treat composite user types like embeddables which would handle many possibly problematic cases like joins, property expansion etc. out of the box.

Yes I think that is definitely the way to go, but I wasn't exactly sure how to achieve it. Are nested properties from embeddables currently reflected in the metamodel?

**Comment by beikov** [comment]
> Yes I think that is definitely the way to go, but I wasn't exactly sure how to achieve it. Are nested properties from embeddables currently reflected in the metamodel?

Yes, an embeddable is completely covered by the metamodel. Maybe we could create custom EmbeddableType instances for composite types?

**Comment by JWGmeligMeyling** [comment]
Interesting hook, I'll do some tests.

**Comment by JWGmeligMeyling** [comment]
Where do you think we should represent this in the metamodel? I believe we quite heavily rely on the standard JPA metamodel over our extended metamodel? Where do you think this could be implemented the easiest? Any specific pointers to occurences of a method to look out for?

**Comment by beikov** [comment]
It would be great if we could completely rely on the extendend metamodel we defined. I'm not sure how much work that is going to be though. I'd be ok with handling composite types specially, like you do now in the PR, given that you handle all the cases where embeddables are treated specially.

**Comment by JWGmeligMeyling** [comment]
I'm closing this as @AFbenHassine won't finish it (he's heading to London for a 2nd masters). I am planning to finish it someday, but it will require quite some work, especially since we want to try the metamodel route. I want to concentrate on my other PRs first (#744 and #746). 

I remembered looking into this briefly when you mentioned it a while back. It seemed going to be quite involved, because a lot of places ([`JpaUtils`](https://github.com/Blazebit/blaze-persistence/blob/master/core/impl/src/main/java/com/blazebit/persistence/impl/JpaUtils.java) in particular) we still use the plain JPA metamodel rather than our own extended metamodel. So everything related to embeddables will have to be changed so that it uses our own metamodel.

**Pull request closed without merging**